src/NuGet.Core/NuGet.Protocol/Plugins/PluginManager.cs

```diff
@@ -150,60 +159,109 @@ public void Dispose()
-
-                pluginCreationResult = new PluginCreationResult(
-                    plugin,
-                    utilities.Value,
-                    lazyOperationClaims.Value.Result);
-            }
-            else
-            {
-                pluginCreationResult = new PluginCreationResult(result.Message);
-            }
+            return await ConcurrencyUtilities.ExecuteWithFileLockedAsync(
+                cacheEntry.CacheFileName,
+                action: async lockedToken =>
+                {
+                    cacheEntry.LoadFromFile();
+                    if (cacheEntry.OperationClaims == null || cacheEntry.OperationClaims.Contains(requestedOperationClaim))
+                    {
+                        if (result.PluginFile.State.Value == PluginFileState.Valid)
+                        {
+                            var plugin = await _pluginFactory.GetOrCreateAsync(
+                                result.PluginFile.Path,
+                                PluginConstants.PluginArguments,
+                                new RequestHandlers(),
+                                _connectionOptions,
+                                cancellationToken);
+
+                            var utilities = await PerformOneTimePluginInitializationAsync(plugin, cancellationToken);
+
+                            // We still make the GetOperationClaims call even if we have the operation claims cached. This is a way to self-update the cache.
+                            var operationClaims = await _pluginOperationClaims.GetOrAdd(
+                                   requestKey,
+                                   key => new Lazy<Task<IReadOnlyList<OperationClaim>>>(() =>
+                                   GetPluginOperationClaimsAsync(
+                                       plugin,
+                                       packageSourceRepository,
+                                       serviceIndex,
+                                       cancellationToken))).Value;
+
+                            if (!EqualityUtility.SequenceEqualWithNullCheck(operationClaims, cacheEntry.OperationClaims))
+                            {
+                                cacheEntry.AddOrUpdateOperationClaims(operationClaims);
+                                await cacheEntry.UpdateCacheFileAsync();
+                            }
+
+                            pluginCreationResult = new PluginCreationResult(
+                                plugin,
+                                utilities.Value,
+                                operationClaims);
+                        }
+                        else
+                        {
+                            pluginCreationResult = new PluginCreationResult(result.Message);
+                        }
+                    }
 
-            return pluginCreationResult;
+                    // Return a dummy uninitialized result if the plugin was not started because the supported operation is not available.
```

Comment:
I wasn't able to understand this dummy uninitialized plugin result. May be some example would make it more clear?

Reply:
The reason why I return a dummy uninitialized plugin is because it's a bit cleaner to handle for the caller.

In the plugin resource provider case,
https://github.com/NuGet/NuGet.Client/blob/10b5c2128351c32f2a2ee09e457fa452082b2103/src/NuGet.Core/NuGet.Protocol/Resources/PluginResource.cs#L78
the caller just asks for its claim and gets nothing and then proceeds. 

In the authentication plugin case, https://github.com/NuGet/NuGet.Client/blob/10b5c2128351c32f2a2ee09e457fa452082b2103/src/NuGet.Clients/NuGet.Credentials/SecurePluginCredentialProvider.cs#L83-L93, same we check for the operation claims and then ignore if it doesn't contain the required one. 

I could probably clean it up, by return a null there, but that's changed the contract that the plugin creation always returns something, and that the caller only needs to check the message for failures. 
If there's a message in the plugin creation result that means the plugin creation has failed. 

In this case, we're not doing it for perf reasons.